server/src/main/java/org/opensearch/index/store/remote/utils/TransferManager.java

```diff
@@ -50,63 +48,115 @@ public TransferManager(final BlobContainer blobContainer, final FileCache fileCa
-                if (cachedIndexInput == null) {
-                    try {
-                        return new FileCachedIndexInput(fileCache, blobFetchRequest.getFilePath(), downloadBlockLocally(blobFetchRequest));
-                    } catch (IOException e) {
-                        logger.warn("Failed to download " + blobFetchRequest.getFilePath(), e);
-                        return null;
+        return AccessController.doPrivileged((PrivilegedAction<FileCachedIndexInput>) () -> {
+            try {
+                if (Files.exists(request.getFilePath()) == false) {
+                    try (
+                        InputStream snapshotFileInputStream = blobContainer.readBlob(
+                            request.getBlobName(),
+                            request.getPosition(),
+                            request.getLength()
+                        );
+                        OutputStream fileOutputStream = Files.newOutputStream(request.getFilePath());
+                        OutputStream localFileOutputStream = new BufferedOutputStream(fileOutputStream)
+                    ) {
+                        snapshotFileInputStream.transferTo(localFileOutputStream);
                     }
-                } else {
-                    if (cachedIndexInput.isClosed()) {
-                        // if it's already in the file cache, but closed, open it and replace the original one
-                        try {
-                            final IndexInput luceneIndexInput = blobFetchRequest.getDirectory()
-                                .openInput(blobFetchRequest.getFileName(), IOContext.READ);
-                            return new FileCachedIndexInput(fileCache, blobFetchRequest.getFilePath(), luceneIndexInput);
-                        } catch (IOException e) {
-                            logger.warn("Failed to open existing file for " + blobFetchRequest.getFilePath(), e);
-                            return null;
-                        }
-                    }
-                    // already in the cache and ready to be used (open)
-                    return cachedIndexInput;
                 }
-            })
-        );
+                final IndexInput luceneIndexInput = request.getDirectory().openInput(request.getFileName(), IOContext.READ);
+                return new FileCachedIndexInput(fileCache, request.getFilePath(), luceneIndexInput);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        });
+    }
+
+    /**
+     * Implementation of CachedIndexInput the defers creation of the underlying
+     * IndexInput until the first invocation of {@link #getIndexInput()}. This
+     * class is thread safe, and concurrent calls to {@link #getIndexInput()} will
+     * result in blocking until the initial thread completes the creation of the
+     * IndexInput.
+     */
+    private static class DelayedCreationCachedIndexInput implements CachedIndexInput {
```

Comment:
I believe the `CompletableFuture` (or similar abstraction) would fit better - one of the benefits would be the cancellation propagating: whoever calls the `close()` while the entry is still being downloaded will be blocked. 

Another thing (which is difficult to trace), but it looks there is no `incRef` call on the key associated with `DelayedCreationCachedIndexInput` (please correct me if I am wrong), so the key be could be removed while it is still being downloaded.

Reply:
> I believe the CompletableFuture (or similar abstraction) would fit better - one of the benefits would be the cancellation propagating: whoever calls the close() while the entry is still being downloaded will be blocked.

Can you expand a bit on how this might look?

> so the key be could be removed while it is still being downloaded.

This was a problem before so there are some multithreaded tests in TransferManagerTests designed to find any races here. In this case, there is an implicit `incRef` as a result of adding the `DelayedCreationCachedIndexInput` instance to the cache via the `compute` call so the instance is protected from removal until that reference is decremented.

Reply:
> Can you expand a bit on how this might look?

I mean, with  `CompletableFuture` we could:
 - check if `CachedIndexInput` is ready to use (`isDone` / ...)
 - ask for transfer to terminate although this is not supported yet (`cancel()`)
 - check and propagate exceptions if any

Reply:
Ugh, I was really over-thinking this. A CompleteableFuture with a simple AtomicBoolean to ensure only one thread starts the work is all I need. The CompleteableFuture handles the "result or error" and we don't need any extra synchronization with close().